File: jar/src/main/java/jline/lib/kpctoolbox/mvph/MVPH.java
/**
 * @file Multivariate Phase-Type Distribution (MVPH) functions
 *
 * Ported from MATLAB: matlab/lib/kpctoolbox/mvph/
 *
 * @since LINE 3.0
 */
package jline.lib.kpctoolbox.mvph;

import org.apache.commons.math3.linear.LUDecomposition;
import org.apache.commons.math3.linear.MatrixUtils;
import org.apache.commons.math3.linear.RealMatrix;

import jline.util.matrix.Matrix;

public final class MVPH {
    private MVPH() {}

    /**
     * Computes the joint moment E[X^n1 * Y^n2] of a bivariate phase-type distribution.
     */
    public static double mvph_joint(double[] alpha, Matrix S, Matrix T, Matrix D, int n1, int n2) {
        int sSize = S.getNumRows();
        int tSize = T.getNumRows();

        RealMatrix negS = MatrixUtils.createRealMatrix(sSize, sSize);
        for (int i = 0; i < sSize; i++) {
            for (int j = 0; j < sSize; j++) {
                negS.setEntry(i, j, -S.get(i, j));
            }
        }

        RealMatrix invNegS = new LUDecomposition(negS).getSolver().getInverse();

        RealMatrix invNegSPow = MatrixUtils.createRealIdentityMatrix(sSize);
        for (int p = 0; p < n1 + 1; p++) {
            invNegSPow = invNegSPow.multiply(invNegS);
        }

        RealMatrix negT = MatrixUtils.createRealMatrix(tSize, tSize);
        for (int i = 0; i < tSize; i++) {
            for (int j = 0; j < tSize; j++) {
                negT.setEntry(i, j, -T.get(i, j));
            }
        }

        RealMatrix invNegT = new LUDecomposition(negT).getSolver().getInverse();

        RealMatrix invNegTPow = MatrixUtils.createRealIdentityMatrix(tSize);
        for (int p = 0; p < n2 + 1; p++) {
            invNegTPow = invNegTPow.multiply(invNegT);
        }

        RealMatrix dMatrix = MatrixUtils.createRealMatrix(sSize, tSize);
        for (int i = 0; i < sSize; i++) {
            for (int j = 0; j < tSize; j++) {
                dMatrix.setEntry(i, j, D.get(i, j));
            }
        }

        RealMatrix minusT = MatrixUtils.createRealMatrix(tSize, tSize);
        for (int i = 0; i < tSize; i++) {
            for (int j = 0; j < tSize; j++) {
                minusT.setEntry(i, j, -T.get(i, j));
            }
        }

        RealMatrix alphaVec = MatrixUtils.createRowRealMatrix(alpha);
        RealMatrix step1 = alphaVec.multiply(invNegSPow);
        RealMatrix step2 = step1.multiply(dMatrix);
        RealMatrix step3 = step2.multiply(invNegTPow);
        RealMatrix step4 = step3.multiply(minusT);

        double[] onesArr = new double[tSize];
        for (int i = 0; i < tSize; i++) onesArr[i] = 1.0;
        RealMatrix ones = MatrixUtils.createColumnRealMatrix(onesArr);
        RealMatrix step5 = step4.multiply(ones);

        double result = step5.getEntry(0, 0);

        return factorial(n1) * factorial(n2) * result;
    }

    public static double mvph_mean_x(double[] alpha, Matrix S, Matrix T, Matrix D) {
        return mvph_joint(alpha, S, T, D, 1, 0);
    }

    public static double mvph_mean_y(double[] alpha, Matrix S, Matrix T, Matrix D) {
        return mvph_joint(alpha, S, T, D, 0, 1);
    }

    public static double mvph_cov(double[] alpha, Matrix S, Matrix T, Matrix D) {
        double exy = mvph_joint(alpha, S, T, D, 1, 1);
        double ex = mvph_joint(alpha, S, T, D, 1, 0);
        double ey = mvph_joint(alpha, S, T, D, 0, 1);
        return exy - ex * ey;
    }

    public static double mvph_corr(double[] alpha, Matrix S, Matrix T, Matrix D) {
        double ex = mvph_joint(alpha, S, T, D, 1, 0);
        double ey = mvph_joint(alpha, S, T, D, 0, 1);
        double ex2 = mvph_joint(alpha, S, T, D, 2, 0);
        double ey2 = mvph_joint(alpha, S, T, D, 0, 2);
        double exy = mvph_joint(alpha, S, T, D, 1, 1);

        double varX = ex2 - ex * ex;
        double varY = ey2 - ey * ey;
        double covXY = exy - ex * ey;

        if (varX <= 0 || varY <= 0) {
            return 0.0;
        }

        return covXY / (Math.sqrt(varX) * Math.sqrt(varY));
    }

    private static double factorial(int n) {
        if (n <= 1) return 1.0;
        double result = 1.0;
        for (int i = 2; i <= n; i++) {
            result *= (double) i;
        }
        return result;
    }
}

File: jar/src/main/kotlin/jline/lib/kpctoolbox/mvph/MVPH.kt
package jline.lib.kpctoolbox.mvph

import jline.util.matrix.Matrix
import org.apache.commons.math3.linear.LUDecomposition
import org.apache.commons.math3.linear.MatrixUtils

/**
 * Multivariate Phase-Type Distribution (MVPH) functions.
 *
 * Ported from MATLAB: matlab/lib/kpctoolbox/mvph/
 */

/**
 * Computes the joint moment E[X^n1 * Y^n2] of a bivariate phase-type distribution.
 *
 * The bivariate phase-type distribution has:
 * - Initial vector alpha
 * - First phase transition matrix S (for X)
 * - Transition matrix D between phases (from X completion to Y start)
 * - Second phase transition matrix T (for Y)
 *
 * Formula: E[X^n1 * Y^n2] = n1! * n2! * alpha * inv(-S)^(n1+1) * D * inv(-T)^(n2+1) * (-T) * e
 *
 * @param alpha Initial probability vector
 * @param S First phase generator matrix (for X)
 * @param T Second phase generator matrix (for Y)
 * @param D Transition matrix between phases
 * @param n1 Power for first variable X
 * @param n2 Power for second variable Y
 * @return Joint moment E[X^n1 * Y^n2]
 */
fun mvph_joint(
    alpha: DoubleArray,
    S: Matrix,
    T: Matrix,
    D: Matrix,
    n1: Int,
    n2: Int
): Double {
    val sSize = S.numRows
    val tSize = T.numRows

    // Convert S to Apache Commons Math format and compute inv(-S)
    val negS = MatrixUtils.createRealMatrix(sSize, sSize)
    for (i in 0 until sSize) {
        for (j in 0 until sSize) {
            negS.setEntry(i, j, -S.get(i, j))
        }
    }

    // Compute inv(-S)
    val invNegS = LUDecomposition(negS).solver.inverse

    // Compute inv(-S)^(n1+1)
    var invNegSPow = MatrixUtils.createRealIdentityMatrix(sSize)
    for (p in 0 until n1 + 1) {
        invNegSPow = invNegSPow.multiply(invNegS)
    }

    // Convert T to Apache Commons Math format and compute inv(-T)
    val negT = MatrixUtils.createRealMatrix(tSize, tSize)
    for (i in 0 until tSize) {
        for (j in 0 until tSize) {
            negT.setEntry(i, j, -T.get(i, j))
        }
    }

    // Compute inv(-T)
    val invNegT = LUDecomposition(negT).solver.inverse

    // Compute inv(-T)^(n2+1)
    var invNegTPow = MatrixUtils.createRealIdentityMatrix(tSize)
    for (p in 0 until n2 + 1) {
        invNegTPow = invNegTPow.multiply(invNegT)
    }

    // Convert D to Apache Commons Math format
    val dMatrix = MatrixUtils.createRealMatrix(sSize, tSize)
    for (i in 0 until sSize) {
        for (j in 0 until tSize) {
            dMatrix.setEntry(i, j, D.get(i, j))
        }
    }

    // Compute (-T) - note: negT is already -T
    // But we need the actual -T matrix for the final multiplication
    val minusT = MatrixUtils.createRealMatrix(tSize, tSize)
    for (i in 0 until tSize) {
        for (j in 0 until tSize) {
            minusT.setEntry(i, j, -T.get(i, j))
        }
    }

    // Compute the chain: alpha * inv(-S)^(n1+1) * D * inv(-T)^(n2+1) * (-T) * e

    // Step 1: alpha * inv(-S)^(n1+1)
    val alphaVec = MatrixUtils.createRowRealMatrix(alpha)
    val step1 = alphaVec.multiply(invNegSPow)

    // Step 2: step1 * D
    val step2 = step1.multiply(dMatrix)

    // Step 3: step2 * inv(-T)^(n2+1)
    val step3 = step2.multiply(invNegTPow)

    // Step 4: step3 * (-T)
    val step4 = step3.multiply(minusT)

    // Step 5: step4 * e (column vector of ones)
    val ones = MatrixUtils.createColumnRealMatrix(DoubleArray(tSize) { 1.0 })
    val step5 = step4.multiply(ones)

    // Get the scalar result
    val result = step5.getEntry(0, 0)

    // Multiply by factorials
    return factorial(n1) * factorial(n2) * result
}

/**
 * Computes the mean of the first variable in a bivariate PH distribution.
 * E[X] = mvph_joint(alpha, S, T, D, 1, 0)
 */
fun mvph_mean_x(alpha: DoubleArray, S: Matrix, T: Matrix, D: Matrix): Double {
    return mvph_joint(alpha, S, T, D, 1, 0)
}

/**
 * Computes the mean of the second variable in a bivariate PH distribution.
 * E[Y] = mvph_joint(alpha, S, T, D, 0, 1)
 */
fun mvph_mean_y(alpha: DoubleArray, S: Matrix, T: Matrix, D: Matrix): Double {
    return mvph_joint(alpha, S, T, D, 0, 1)
}

/**
 * Computes the covariance of a bivariate PH distribution.
 * Cov(X, Y) = E[XY] - E[X]*E[Y]
 */
fun mvph_cov(alpha: DoubleArray, S: Matrix, T: Matrix, D: Matrix): Double {
    val exy = mvph_joint(alpha, S, T, D, 1, 1)
    val ex = mvph_joint(alpha, S, T, D, 1, 0)
    val ey = mvph_joint(alpha, S, T, D, 0, 1)
    return exy - ex * ey
}

/**
 * Computes the correlation of a bivariate PH distribution.
 * Corr(X, Y) = Cov(X, Y) / (StdDev(X) * StdDev(Y))
 */
fun mvph_corr(alpha: DoubleArray, S: Matrix, T: Matrix, D: Matrix): Double {
    val ex = mvph_joint(alpha, S, T, D, 1, 0)
    val ey = mvph_joint(alpha, S, T, D, 0, 1)
    val ex2 = mvph_joint(alpha, S, T, D, 2, 0)
    val ey2 = mvph_joint(alpha, S, T, D, 0, 2)
    val exy = mvph_joint(alpha, S, T, D, 1, 1)

    val varX = ex2 - ex * ex
    val varY = ey2 - ey * ey
    val covXY = exy - ex * ey

    if (varX <= 0 || varY <= 0) {
        return 0.0
    }

    return covXY / (kotlin.math.sqrt(varX) * kotlin.math.sqrt(varY))
}

/**
 * Computes factorial.
 */
private fun factorial(n: Int): Double {
    if (n <= 1) return 1.0
    var result = 1.0
    for (i in 2..n) {
        result *= i.toDouble()
    }
    return result
}
